### PR TITLE
Rerender the sticky scroll upon change of model tokens (fixes #156266)

### LINKED ISSUE #156266: Sticky scroll: lines not colored after window reload
Testing #156175
Recording 2022-07-26 at 11 58 38

### PR DESCRIPTION
**Pull request by aiday-mar**
Re-renders the sticky scroll widget on the change of model tokens when the line range of the change intersects the line range of the widget (fixes #156266)

### CODE REVIEW

File: src/vs/editor/contrib/stickyScroll/browser/stickyScroll.ts [hunk 1]
```diff
@@ -66,12 +67,24 @@ class StickyScrollController extends Disposable implements IEditorContribution {
...
 	}
 
+	private _onTokensChange(event: IModelTokensChangedEvent) {
+		const beginningLine = this._editor.getVisibleRanges()[0].getStartPosition().lineNumber - 1;
+		const endLine = beginningLine + this.stickyScrollWidget.codeLineCount + 1;
+		const fromLineNumber = event.ranges[0].fromLineNumber;
+		const toLineNumber = event.ranges[0].toLineNumber;
+		if (fromLineNumber > endLine || beginningLine > toLineNumber) {
+			return;
+		}
```
Comment (alexdima) [inline src/vs/editor/contrib/stickyScroll/browser/stickyScroll.ts:84]: Here we should intersect all ranges present in `event.ranges`, not just `event.ranges[0]` with all the currently rendered `StickyScrollCodeLine`. 

The `StickyScrollWidget` could have a getter like `.getCurrentStickyLineNumbers(): number[]` which would return the line numbers currently rendered at the top and then that array would be intersected with each of the ranges in `event.ranges`

**Review by alexdima — CHANGES_REQUESTED** [review]

File: src/vs/editor/contrib/stickyScroll/browser/stickyScroll.ts (hunk 1 above)
Comment (aiday-mar) [reply src/vs/editor/contrib/stickyScroll/browser/stickyScroll.ts:84]: Right, I'll change that

File: src/vs/editor/contrib/stickyScroll/browser/stickyScroll.ts [hunk 2]
```diff
@@ -335,6 +355,7 @@ class StickyScrollCodeLine {
 class StickyScrollWidget implements IOverlayWidget {
 
 	private readonly arrayOfCodeLines: StickyScrollCodeLine[] = [];
+	private readonly linesRange: number[] = [];
```
Comment (alexdima) [inline src/vs/editor/contrib/stickyScroll/browser/stickyScroll.ts:358]: Adding a member here is not necessary. Each `StickyScrollCodeLine` knows its line number, it just needs to expose a getter or make it public. See `StickyScrollCodeLine._lineNumber` which already has this.

File: src/vs/editor/contrib/stickyScroll/browser/stickyScroll.ts [hunk 3]
```diff
@@ -347,7 +368,15 @@ class StickyScrollWidget implements IOverlayWidget {
 		return this.arrayOfCodeLines.length;
 	}
 
-	pushCodeLine(codeLine: StickyScrollCodeLine) {
+	getCurrentLineRange(): number[] {
```
Comment (alexdima) [inline src/vs/editor/contrib/stickyScroll/browser/stickyScroll.ts:371]: This should return a set of rendered lines, not just the extremes.

File: src/vs/editor/contrib/stickyScroll/browser/stickyScroll.ts [hunk 4]
```diff
@@ -66,12 +67,31 @@ class StickyScrollController extends Disposable implements IEditorContribution {
...
+		let rerender: boolean = false;
+		for (const range of event.ranges) {
+			const fromLineNumber = range.fromLineNumber;
+			const toLineNumber = range.toLineNumber;
+			const fromLineWidget = widgetLineRange[0];
+			const toLineWidget = widgetLineRange[1];
+			if (fromLineNumber <= toLineWidget && toLineNumber >= fromLineWidget) {
+				rerender = true;
+				break;
+			}
```
Comment (alexdima) [inline src/vs/editor/contrib/stickyScroll/browser/stickyScroll.ts:88]: This should handle that `widgetLineRange` is a set of N lines, not just the extremes.

**Review by alexdima — CHANGES_REQUESTED** [review]
I'm sorry I think my example might have misled you because it only contained 2 sticky lines. Here's an updated example:

image

`StickyScrollWidget.getCurrentLines()` would return in this case `[111, 272, 351]` and these line numbers would need to be checked if they are in a range i.e.:

```ts

private _affectsStickyLineNumber(event: IModelTokensChangedEvent) {
	const stickyLineNumbers = this.stickyScrollWidget.getCurrentLines();
	for (const stickyLineNumber of stickyLineNumbers) {
		for (const range of event.ranges) {
			if (stickyLineNumber >= range.fromLineNumber && stickyLineNumber <= range.toLineNumber) {
				return true;
			}
		}
	}
	return false;
}

private _onTokensChange(event: IModelTokensChangedEvent) {
	if (this._affectsStickyLineNumber(event)) {
		this._update(false);
	}
}
```

**Review by alexdima — APPROVED** [review]
Nice!

**Pull request merged**